osu.Game/Screens/Play/Player.cs

```diff
@@ -281,6 +284,12 @@ private void load(AudioManager audio, OsuConfigManager config, OsuGameBase game)
                 ScoreProcessor.RevertResult(r);
             };
 
+            DimmableStoryboard.HasStoryboardEnded.ValueChanged += _ =>
+            {
+                if (ScoreProcessor.HasCompleted.Value)
+                    scheduleCompletion();
```

Comment:
Guard checking all completion conditions at the start of `updateCompletionState()` sounds fine.

Reply:
How should we cause the player to fail?

Reply:
Make a private custom mod that causes the player to fail immediately and not restart on fail (check `IApplicableFailOverride`).

Reply:
How do we specify failure before the first hit circle?

Reply:
Hmm, that's actually not a good solution, disregard it. Instead: 
 1. override `HasCustomSteps` to true, and call `CreateTest(null)` on all test cases. (that needs to happen any way you slice it)
 2. make a private field in the test scene holding the test case's own fail conditions, reset it on `SetUp` and set it on the test case you want the player to fail on, by: 
 ```csharp
 CreateTest(() => 
 {
     AddStep("fail on first judgement", () => failConditions = (_, __) => true); 
 }
 ```
 
 3. then pass that fail conditions field to `OutroPlayer` and subscribe to `HealthProcessor.FailConditions` with the given conditions.

Reply:
Why is it bad to just alter the `FailConditions` on one testcase? 

We found that we actually don't need to fail immediately- we're having trouble getting the storyboard to end during the fail animation while still having time for the skip overlay to be clicked in `TestSkipOutro`. We want to change the test to check if the overlay is simply visible instead, but the `skipOutroOverlay` is private so we can't access it with `State.Value`. Is there some way to check if some overlay is visible?

Reply:
> Why is it bad to just alter the `FailConditions` on one testcase?
>
> We found that we actually don't need to fail immediately-

I'm not sure what you mean, do you have an alternative to what I proposed? if so, post it here.

> we're having trouble getting the storyboard to end during the fail animation while still having time for the skip overlay to be clicked in `TestSkipOutro`. We want to change the test to check if the overlay is simply visible instead, but the `skipOutroOverlay` is private so we can't access it with `State.Value`. Is there some way to check if some overlay is visible?

Usually, it's retrieved by using the testing extension `ChildrenOfType<T>`, but since there are two skip overlays I would say just expose it (`protected` at base and `public` at test class)

Reply:
> I'm not sure what you mean, do you have an alternative to what I proposed? if so, post it here.

I meant we didn't need the player to fail immediately during the test case, in fact doing that would make the storyboard end further after the fail animation (which doesn't test what we want.) So it's fine for the player to fail after the first judgement.

If the storyboard lasts 2000ms, then it won't end during the fail animation. We found that if the storyboard is 1300ms, then it will end during the fail animation but then `TestSkipOutro` is broken because the storyboard ends too quickly after the last hitobject for the test to click on the skip overlay though it does show up briefly. 

> Usually, it's retrieved by using the testing extension `ChildrenOfType<T>`, but since there are two skip overlays I would say just expose it (`protected` at base and `public` at test class)

It seems like I still can't access `State.Value` because it's inside a private class `FadeContainer`. Should I use `ChildrenOfType<T>`?

Reply:
Checking if the skip overlay is present seems to work, but then it can't test if clicking on it shows the score.

Is there some way I could change the storyboard duration before each test? Otherwise it seems like I'll have to make a new test scene for the fail overlay.

Reply:
As I said, you'll have to manually load the player for each test case to achieve that, check the steps in https://github.com/ppy/osu/pull/12432#discussion_r615468528, add another field `currentStoryboard` and set it in the same place as the fail conditions:
```csharp
CreateTest(() =>
{
    AddStep("set storyboard" () => currentStoryboard = ...); 
    AddStep("fail on first judgement", () => currentFailConditions = (_, __) => true); 
});
```

then pass `currentStoryboard` in `CreateWorkingBeatmap` and there you have it.